vaultclient/iam_auth.go

```diff
@@ -0,0 +1,130 @@
+package vaultclient
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/hashicorp/vault/api"
+	"github.com/hootsuite/vault-ctrl-tool/util"
+	jww "github.com/spf13/jwalterweatherman"
+)
+
+func stsSigningResolver(service, region string, optFns ...func(*endpoints.Options)) (endpoints.ResolvedEndpoint, error) {
+	defaultEndpoint, err := endpoints.DefaultResolver().EndpointFor(service, region, optFns...)
+	if err != nil {
+		return defaultEndpoint, err
+	}
+
+	defaultEndpoint.SigningRegion = region
+	return defaultEndpoint, nil
+}
+
+func generateLoginData(creds *credentials.Credentials, configuredRegion string) (map[string]interface{}, error) {
+	loginData := make(map[string]interface{})
+
+	stsSession, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Credentials:      creds,
+			Region:           &configuredRegion,
+			EndpointResolver: endpoints.ResolverFunc(stsSigningResolver),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var params *sts.GetCallerIdentityInput
+	svc := sts.New(stsSession)
+	stsRequest, _ := svc.GetCallerIdentityRequest(params)
+
+	stsRequest.Sign()
+
+	// Now extract out the relevant parts of the request
+	headersJson, err := json.Marshal(stsRequest.HTTPRequest.Header)
+	if err != nil {
+		return nil, err
+	}
+	requestBody, err := ioutil.ReadAll(stsRequest.HTTPRequest.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	loginData["iam_http_request_method"] = stsRequest.HTTPRequest.Method
+	loginData["iam_request_url"] = base64.StdEncoding.EncodeToString([]byte(stsRequest.HTTPRequest.URL.String()))
+	loginData["iam_request_headers"] = base64.StdEncoding.EncodeToString(headersJson)
+	loginData["iam_request_body"] = base64.StdEncoding.EncodeToString(requestBody)
+
+	return loginData, nil
+}
+
+func getSecret(c *api.Client, creds *credentials.Credentials) (*api.Secret, error) {
+
+	loginData, err := generateLoginData(creds, "us-east-1")
+	if err != nil {
+		return nil, err
+	}
+	if loginData == nil {
+		return nil, fmt.Errorf("got nil response from generateLoginData")
+	}
+
+	loginData["role"] = util.Flags.IamAuthRole
+
+	secret, err := c.Logical().Write("auth/aws/login", loginData)
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil {
+		return nil, fmt.Errorf("empty response from credential provider")
+	}
+
+	return secret, nil
+}
+
+func getCredentialsFromRole() (*credentials.Credentials, error) {
+	roleProvider := &ec2rolecreds.EC2RoleProvider{
+		Client: ec2metadata.New(session.New()),
```

Comment:
New is deprecated. Please see its recommendations, or if you can't use it, please file a ticket against the Go AWS SDK and link to it in the code as a comment.